File: src/main/scala/net/liftweb/seventhings/snippet/ShowCode.scala
package net.liftweb.seventhings
package snippet

import net.liftweb._
import http._
import util._
import Helpers._
import common._

import scala.xml.NodeSeq

/**
 * A snippet that will show part or all of a source file
 * that's part of this project
 */
object ShowCode {
  def render: NodeSeq =
    for {
      fileName <- S.attr("name") ?~ "Name missing"

      code <- (LiftRules.loadResource("/src/scala"+fileName) or
               LiftRules.loadResource("/src/webapp"+fileName)) ?~ 
        ("Cannot find "+fileName)

    } yield {
      def calcBrush = if (fileName.endsWith(".html") ||
                          fileName.endsWith(".xml")) "html" else "scala"

      def gitDir = if (fileName.endsWith(".html") ||
                          fileName.endsWith(".xml")) "webapp" else "scala"

      val start = S.attr("start")
      val end = S.attr("end")

      val rawCode = (new String(code, "UTF-8").
                     split("""\n""").
                     toList.map(_.replace("\t", "        "))).
      dropWhile(s => start.map(st => s.indexOf(st) < 0) openOr false) match {
        case Nil => Nil
        case x :: xs => x :: xs.
        takeWhile(s => end.map(en => s.indexOf(en) < 0) openOr true)
      }


      

      <div style="text-align: center"><tt>Listing:
      <a href={"https://github.com/lift/seventhings/tree/master/src/main/"+
               gitDir+fileName}>{fileName}</a></tt>
      <pre class={"listing brush: "+calcBrush}>{rawCode.mkString("\n")}</pre>
      </div>
    }
    
  private implicit def boxNStoNS(in: Box[NodeSeq]): NodeSeq =
    in match {
      case Full(ns) => ns
      case Failure(msg, _, _) => 
        <div style="display: block; margin: 8px; border: 2px solid red">
      Failed to look up source: {msg}
      </div>
      case _ => <div style="display: block; margin: 8px; border: 2px solid red">
      Failed to look up source
      </div>


    }
}

File: src/main/java/net/liftweb/seventhings/snippet/ShowCode.java
package net.liftweb.seventhings.snippet;


import net.liftweb.common.Box;
import net.liftweb.http.LiftRulesJ;
import net.liftweb.http.SJ;
import scala.Option;
import scala.xml.Elem;
import scala.xml.NodeSeq;
import scala.xml.Text;

import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.List;

import static net.liftweb.seventhings.lib.XmlJ.*;

public class ShowCode {

    private String calcBrush(String fileName) {
        if (fileName.endsWith(".html") || fileName.endsWith(".xml")) {
            return "html";
        }

        if (fileName.endsWith(".java")) {
            return "scala";
        }

        return "scala";
    }

    private String gitDir(String fileName) {
        if (fileName.endsWith(".html") || fileName.endsWith(".xml")) {
            return "webapp";
        }

        if (fileName.endsWith(".java")) {
            return "java";
        }

        return "scala";
    }


    public NodeSeq render() {

        String codeListing = "";

        Option<String> fileNameO = SJ.j().attrsFlattenToMap().get("name");
        if (fileNameO.isDefined()) {

            String fileName = fileNameO.get();

//            System.out.println("fileName="+fileName);

            // does not work with .or(LiftRulesJ.j().loadResource("/src/webapp"+fileName))
            Box<byte[]> codeBox = (LiftRulesJ.j().loadResource("/src/java"+fileName)).failMsg("Cannot find "+fileName);

            if (!codeBox.isDefined()) {
                codeBox = (LiftRulesJ.j().loadResource("/src/webapp"+fileName)).failMsg("Cannot find "+fileName);
            }

            if (codeBox.isDefined()) {

                byte[] code = codeBox.openTheBox();

                Option<String> startO = SJ.j().attrsFlattenToMap().get("start");
                Option<String> endO = SJ.j().attrsFlattenToMap().get("end");

                // Option.getOrElse did not compile, so this dirty workaround
                String start = "";
                String end = "";

                if (startO.isDefined()) { start = startO.get(); }
                if (endO.isDefined())   { end = endO.get(); }

                codeListing = formatCodeListing(code, start, end );

//                System.out.println("CODE: "+codeListing);

                Elem el = elem("div", attr("style", "text-align: center"),
                        elem("tt",
                            text("Listing:"),
                            elem("a", attr("href", "https://github.com/lift/seventhings/tree/liftj/src/main/" + gitDir(fileName) + fileName),
                                text(fileName))),
                        elem("pre", attr("class", "listing brush: "+calcBrush(fileName)), text(codeListing)));

                return toNodeSeq(el);
            }
        }
        return scala.xml.NodeSeq.seqToNodeSeq(new Text("Failure").theSeq());

    }


    private String formatCodeListing(byte[] code, String start, String end) {

        boolean startDefined = (start == null || start.isEmpty()) ? false : true;
        boolean startFound   = false;
        boolean endDefined   = (end == null || end.isEmpty()) ? false : true;
        boolean endFound     = false;
        boolean firstLineAfterStartRead = startDefined ? false : true;

        try {
            String[] rawCodeArr = new String(code, "UTF-8").split("\\n");
            List<String> rawCode = new ArrayList<String>();
            for (String s : rawCodeArr) {
                String s2 = s.replace("\t", "        ");
                if (startDefined && !startFound) {
                    if (s2.indexOf(start) > -1) {
                        startFound  = true;
                    } else {
                        continue;
                    }
                }
                if (endDefined && firstLineAfterStartRead) {
                    if (!endFound) {
                        if (s2.indexOf(end) > -1) {
                            endFound = true;
                            continue;
                        }
                    } else {
                        continue;
                    }
                }
                rawCode.add(s2);

                if (startFound) {
                    firstLineAfterStartRead = true;
                }
            }

            StringBuilder sb = new StringBuilder();
            for (String s : rawCode) {
                sb.append(s).append("\n");
            }

            return sb.toString();

        } catch (UnsupportedEncodingException e) {
            throw new IllegalArgumentException(e);
        }

    }

}
